### PR TITLE
Issue/refactor add media to editor functionality

### LINKED ISSUE #10669: Rewrite AddMediaToEditor task
Goal of this task is to extract [this task](https://github.com/wordpress-mobile/WordPress-Android/blob/f89f32f0bfa27b8e2ac3db4a73b325afffb6f497/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/EditorMedia.kt#L199) into a testable class and rewrite the logic using coroutines.

### PR DESCRIPTION
**Pull request by malinajirka**
Fixes #10669 

**Merge instructions**:
**This PR is based on a branch from https://github.com/wordpress-mobile/WordPress-Android/pull/10664. The target branch should be switched to `issue/master-extract-editor-media`.**

Note:
I'm planning to write tests for the UseCase in https://github.com/wordpress-mobile/WordPress-Android/issues/10670. If you'd prefer to wait for the tests as they might help with the review, let me know.

Goal of this task was to rewrite the logic related to adding media to the editor. The core of the solution is based on the previous implementation, however, most parts were rewritten from scratch. 

The code in EditorMedia and EditPostActivity isn't final and we'll keep refactoring it. However, the code in `OptimizeAndAddMediaToEditorUseCase` should be close to the final code. We'll probably want to remove the `EditorMedia` reference from the UseCase but I wanted to create the PR today and I thought we could do that later.

To test - I'm writing these tests down now, when the code is in my fresh memory. However, we might want to test this feature when we are about to merge `master-extract-editor-media`.

Test these scenarios for - WPCom/Self-hosted sites with Aztec/Gutenberg
1. Open the editor
2. Click on Add media button
3. Add a media item from the local storage
4. Make sure the item is added and uploaded
5. Switch to html mode and verify the media entry contains url and not a local path
------------------------------------------------
1. Open the editor
2. Click on Add media button
3. Select Camera -> Take photo
4.Make sure the item is added and uploaded
5. Switch to html mode and verify the media entry contains url and not local path
--------------------------------------------------
(available only in Aztec - Gutenberg Gallery block hasn't been implemented yet)
1. Open the editor
2. Click on Add media button
3. Select multiple media items from the local storage
4. Make sure all the items are added and uploaded
5. Switch to html mode and verify that all the media item entries contain url and not local paths
------------------------------------------------
Update after Oguz's suggestion

1. Do a fresh installation
2. Open the editor
3. Click on Add media button
4. Add a media item from the local storage
5. Make sure "Turn on image optimization" dialog is shown
6. Tap on "Turn on" (maybe we should also test the "Leave off" option)
7. Make sure the item is added and uploaded
8. Switch to html mode and verify the media entry contains url and not a local path

------------------------------------------------

1. Do a fresh installation
2. Open the editor
3. Click on Add media button
4. Add a media item from the local storage
5. Make sure "Turn on image optimization" dialog is shown
6. Tap on "Turn Off" 
7. Make sure the item is added and uploaded
8. Switch to html mode and verify the media entry contains url and not a local path

PR submission checklist:

- [x] I have considered adding unit tests where possible. -> tests will be written in https://github.com/wordpress-mobile/WordPress-Android/issues/10670

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    fun optimizeAndAddAsync(
+        uriList: List<Uri>,
+        site: SiteModel,
+        isNew: Boolean,
+        editorMediaListener: EditorMediaListener,
+        editorMedia: EditorMedia
+    ) {
+        launch {
+            _uiState.value = if (uriList.size > 1) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: Should this be a `postValue` call instead since it's in a bg thread?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            )
+            addMediaToEditor(editorMediaListener, enqueuedFiles)
+
+            editorMediaListener.savePostAsyncFromEditorMedia()
+
+            if (optimizeMediaResult.someMediaCouldNotBeRetrieved) {
+                _snackBarMessage.value = SnackbarMessageHolder(R.string.gallery_error)
+            }
+
+            _uiState.value = AddingMediaIdle
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: Similar to my comment, should this be a `postValue` call instead since it's in a bg thread?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    private suspend fun optimizeMediaAsync(
+        site: SiteModel,
+        isNew: Boolean,
+        uriList: List<Uri>
+    ): OptimizeMediaResult {
+        return withContext(bgDispatcher) {
+            uriList
+                    .map { async { optimizeMediaAndTrackEvent(it, isNew, site) } }
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: Since this whole operation is on a bg thread, is it necessary to have another `async` block here? I think the following would do the same thing:

```
uriList.mapNotNull { optimizeMediaAndTrackEvent(it, isNew, site) }
        .let {
            OptimizeMediaResult(
                    optimizedMediaUris = it,
                    someMediaCouldNotBeRetrieved = it.size < uriList.size
            )
        }

```

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 4]
```diff
@@ -1099,7 +1099,11 @@ private void replaceDrawable(Drawable newDrawable) {
         }
     }
 
-    @Override public void appendMediaFiles(Map<String, MediaFile> mediaList) { }
+    @Override public void appendMediaFiles(Map<String, MediaFile> mediaList) {
```
Comment (oguzkocer) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: @malinajirka Can you let me know why this change was necessary? I couldn't find if this was moved from somewhere or fixing a bug. I just didn't want to skip reviewing this, I am sure it's necessary :)

**Comment by oguzkocer** [comment]
@malinajirka I've tested all the cases you mentioned and all of them passed. I think there is at least one more case we can test:

1. Do a fresh installation
2. Open the editor
3. Click on Add media button
4. Add a media item from the local storage
5. Make sure "Turn on image optimization" dialog is shown
6. Tap on "Turn on" (maybe we should also test the "Leave off" option)
7. Make sure the item is added and uploaded
8. Switch to html mode and verify the media entry contains url and not a local path

**Comment by oguzkocer** [comment]
@malinajirka One more thing, sorry for all the comments. I just fixed the merge conflict that I caused by fixing the lint issue in the base branch. I thought I'd save you the time to do it since I knew where to look. I hope you don't mind!

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt (hunk 1 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: Default thread is actually mainThread - see https://github.com/wordpress-mobile/WordPress-Android/blob/dcd4896da50dc64764136b564dacced69d4fe9f3/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt#L50.


Most of the logic still runs on the UIThread because most of the utility methods access view - eg. toast/snackbar. However, the most heavy work - image optimization - works on a background thread.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt (hunk 2 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: As explained above.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt (hunk 3 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: hmm interesting question - would it run in parallel or one after another?

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java (hunk 4 above)
Comment (malinajirka) [reply libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: Before this PR, we'd call either `appendMediaFiles` or `appendMediaFile` depending on the `allowMultipleSelection`. I removed that flag and decided to always call `appendMediaFiles` even when I have only one media file. However, this method wasn't implemented in AztecEditorFragment -> therefore I added this simple implementation.

**Comment by malinajirka** [comment]
Awesome, thank you for the review Oguz! I've updated the PR description with the test scenarios you suggested - good call!

I've replied to the comments - I'll test if it makes any difference if we call `async` tomorrow as I'm not 100% sure how it works and I'm really curious about it;).

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt (hunk 1 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: 🤦‍♂

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt (hunk 3 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/OptimizeAndAddMediaToEditorUseCase.kt]: I think you're right it'd still be in bg thread but not in parallel. Let's keep this as is. There is some minor modifications we can do to improve it, but not really necessary.
